test/middleware_stack_test.rb

```diff
@@ -35,8 +35,9 @@ def test_allows_rebuilding
   end
 
   def test_allows_extending
-    build_stack Apple
+    build_handlers_stack Apple
     @conn.use Orange
+    @conn.adapter :test, &test_adapter
     assert_handlers %w[Apple Orange]
```

Comment:
Yes this test probably overlooked the fact that `build_stack` always adds the adapter at the end, going against the README itself.
Your change works but I see that `build_stack` and `build_handlers_stack` are practically the same thing (and the latter accepts a block but doesn't use it).
Another, better in my opinion, approach would be to add an optional parameter to `build_stack` to skip the adapter insertion at the end. This should reduce code duplication even further.